Adds WeakPtr<ModelPlayerProvider> to HTMLModelElement for deallocation purposes.

amarchenko opened the pull request:
#### b82b1b09779a98c0183704ec16e01b08518ed5a3
<pre>
Adds WeakPtr&lt;ModelPlayerProvider&gt; to HTMLModelElement for deallocation purposes.
<a href="https://bugs.webkit.org/show_bug.cgi?id=288321">https://bugs.webkit.org/show_bug.cgi?id=288321</a>
<a href="https://rdar.apple.com/145434268">rdar://145434268</a>

Reviewed by Ryosuke Niwa.

ModelPlayerProvider is a gateway to communicate with the Model Process.
When an HTMLModelElement is detached from Page/Document that communication
line is gone. To make sure the Model Process is notified when an element&apos;s
destructor is called, WeakPtr&lt;ModelPlayerProvider&gt; is added.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::m_environmentMapReadyPromise):
(WebCore::HTMLModelElement::~HTMLModelElement):
(WebCore::HTMLModelElement::createModelPlayer):
(WebCore::HTMLModelElement::deleteModelPlayer):
(WebCore::HTMLModelElement::didFinishLoading):
(WebCore::HTMLModelElement::insertedIntoAncestor):
(WebCore::HTMLModelElement::removedFromAncestor):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
Keep track of ModelPlayerProvider when an element attached/detached to/from
document and release resources.

* Source/WebCore/Modules/model-element/ModelPlayerProvider.h:
ModelPlayerProvider is RefCounted and CanMakeWeakPtr now.

* Source/WebCore/Modules/model-element/dummy/DummyModelPlayerProvider.cpp:
(WebCore::DummyModelPlayerProvider::create):
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayerProvider.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebKit/WebProcess/Model/WebModelPlayerProvider.cpp:
(WebKit::WebModelPlayerProvider::create):
* Source/WebKit/WebProcess/Model/WebModelPlayerProvider.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):
UniqueRef is not applicable for RefCounted objects, hence switch to Ref.

Canonical link: <a href="https://commits.webkit.org/291171@main">https://commits.webkit.org/291171@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee5981f93c42d895957e57b11ccfcaae1d8ba80a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91982 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11514 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1060 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96911 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42580 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94032 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11854 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20001 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70574 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28062 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94983 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9034 "Found 1 new test failure: fast/forms/ios/focus-input-in-fixed.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83316 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50902 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/91477 "Build is in progress. Recent messages:") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8764 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/917 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41795 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79085 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/912 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98943 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19097 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14121 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79612 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19349 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79168 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78842 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19564 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23365 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/687 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12135 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19078 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24287 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18775 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22234 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20526 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1137,6 +1142,24 @@ bool HTMLModelElement::isURLAttribute(const Attribute& attribute) const
+{
+    auto insertResult = HTMLElement::insertedIntoAncestor(insertionType, parentOfInsertedTree);
+
+    if (document().page())
```
You need to check `insertionType.connectedToDocument`

amarchenko replied:
Done.

rniwa commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1137,6 +1142,24 @@ bool HTMLModelElement::isURLAttribute(const Attribute& attribute) const
+{
+    HTMLElement::removedFromAncestor(removalType, oldParentOfRemovedTree);
+
+    deleteModelPlayer();
```
You need to check `removalType.disconnectedFromDocument`.

amarchenko replied:
Done.

rniwa commented on Source/WebCore/Modules/model-element/HTMLModelElement.h:
```diff
@@ -214,6 +215,10 @@ class HTMLModelElement final : public HTMLElement, private CachedRawResourceClie
 #endif
     std::optional<PlatformLayerIdentifier> modelContentsLayerID() const final;
 
+    // HTMLelement
```
This comment is not right. These are member functions on Node, not HTMLElement.
I suggest we just remove this comment.

amarchenko replied:
Done.

rniwa commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1137,6 +1142,25 @@ bool HTMLModelElement::isURLAttribute(const Attribute& attribute) const
         || HTMLElement::isURLAttribute(attribute);
 }
 
+// Updating modelPlayerProvider when attached to a document.
```
I don't think this comment is useful. It repeats what the code below conveys. Remove?

amarchenko replied:
Removed.

rniwa commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -321,8 +324,10 @@ void HTMLModelElement::createModelPlayer()
 {
-    if (m_modelPlayer && document().page())
-        document().page()->modelPlayerProvider().deleteModelPlayer(*m_modelPlayer);
+    RefPtr strongModelPlayerProvider = m_modelPlayerProvider.get();
```
Nit: WebKit's naming convention for these variables is to use "protected" prefix.

amarchenko replied:
Addressed.

rniwa reviewed (approved):
r=me with some nits.

amarchenko committed:
Adds WeakPtr<ModelPlayerProvider> to HTMLModelElement for deallocation purposes.
https://bugs.webkit.org/show_bug.cgi?id=288321
rdar://145434268

Reviewed by Ryosuke Niwa.

ModelPlayerProvider is a gateway to communicate with the Model Process.
When an HTMLModelElement is detached from Page/Document that communication
line is gone. To make sure the Model Process is notified when an element's
destructor is called, WeakPtr<ModelPlayerProvider> is added.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::m_environmentMapReadyPromise):
(WebCore::HTMLModelElement::~HTMLModelElement):
(WebCore::HTMLModelElement::createModelPlayer):
(WebCore::HTMLModelElement::deleteModelPlayer):
(WebCore::HTMLModelElement::didFinishLoading):
(WebCore::HTMLModelElement::insertedIntoAncestor):
(WebCore::HTMLModelElement::removedFromAncestor):
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
Keep track of ModelPlayerProvider when an element attached/detached to/from
document and release resources.

* Source/WebCore/Modules/model-element/ModelPlayerProvider.h:
ModelPlayerProvider is RefCounted and CanMakeWeakPtr now.

* Source/WebCore/Modules/model-element/dummy/DummyModelPlayerProvider.cpp:
(WebCore::DummyModelPlayerProvider::create):
* Source/WebCore/Modules/model-element/dummy/DummyModelPlayerProvider.h:
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/PageConfiguration.cpp:
(WebCore::PageConfiguration::PageConfiguration):
* Source/WebCore/page/PageConfiguration.h:
* Source/WebKit/WebProcess/Model/WebModelPlayerProvider.cpp:
(WebKit::WebModelPlayerProvider::create):
* Source/WebKit/WebProcess/Model/WebModelPlayerProvider.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):
UniqueRef is not applicable for RefCounted objects, hence switch to Ref.

Canonical link: https://commits.webkit.org/291171@main
